tinygrad/ops.py

```diff
@@ -163,16 +163,15 @@ def forward(ctx, x, w, stride=1, groups=1):
     rcout = cout//ctx.groups
 
     ctx.save_for_backward(x, w)
-    ret = np.zeros((bs, cout, oy, ox), dtype=w.dtype)
 
-    for g in range(ctx.groups):
-      tw = w[g*rcout:(g*rcout+rcout)].reshape(rcout, -1).T
-      for Y in range(oy):
-        for X in range(ox):
-          iY,iX = Y*ys, X*xs
-          tx = x[:, g*cin:(g*cin+cin), iY:iY+H, iX:iX+W].reshape(bs, -1)
-          ret[:, g*rcout:(g*rcout+rcout), Y, X] += tx.dot(tw)
-    return ret
+    gx = x.reshape(bs,ctx.groups,cin,x.shape[2],x.shape[3])
+    str_x = np.lib.stride_tricks.as_strided(
+                gx,
+                shape=(bs, ctx.groups, cin, oy, ox, H, W),
+                strides=(gx.strides[0], gx.strides[1], gx.strides[2], gx.strides[3]*ys, gx.strides[4]*xs, gx.strides[3], gx.strides[4]),
+                writeable=False, 
+            )
+    return np.einsum('igjYXyx,gkjyx -> igkYX', str_x, w.reshape(ctx.groups,rcout, cin, H, W)).reshape(bs, cout, oy, ox)
```

Comment:
perhaps tw=w.reshape(ctx.groups,rcout, cin, H, W) by the gx line?

Reply:
I had it like this before